### DISCUSSION TITLE
NodeJS TypeScript setup

### DISCUSSION THREAD

**Discussion by jangot** [urql-graphql/urql#2765] [Q&A]
Hello.
I just started to work with URQL and I have a problem with the setup of the client on NodeJS + TS server. 
I have a simple query:
```
import { createClient, defaultExchanges } from '@urql/core';

import config from './config';
import { graphql } from './gql';

// I need to have typing here
export default await function getFeature(): Promise<any> {
    const QUERY = graphql(`
        query features {
                featureFlags {
                    features {
                        namespace
                        feature
                    }
                }
            }
    `);
    return createClient({
        url: config.graphqlUrl,
        exchanges: defaultExchanges,
    }).query(QUERY, {}).toPromise();
};
```
And it looks great to me. But I need to set up typing and I don't understand how should it work. If I use an example from `codegen` I see a strange result without typing to set to result of my function.

I see many examples for react applications but can not find something for NodeJS although in the documentation for the library is information that I can use in NodeJs. It was a reason why I started to use it. And I was confused when I don't see any examples for NodeJS or for vanilla JS.
Please advise me on how can I understand how to use URQL on my server or may be better to find other libraries if you don't support NodeJS.

**Comment by JoviDeCroock** [comment]
Hey and welcome,

You should be able to type your results by means of a generic:

```
type Todo = { id: string };
type TodoInput = { user: String }
query<Todo, TodoInput>
```

This can be automated with something like `TypedDocumentNode` which can infer it from the document https://www.the-guild.dev/graphql/codegen/plugins/typescript/typed-document-node

**Comment by jangot** [comment]
Ok, I think I start to understand. But I still have questions.
I can use types only if I use `TypedDocumentNode`. Right?

And there is no way to get type for part of my response.
For example, if I have this data structure:
```
query user {
    me {
        id
        pet {
            name
         }
    }
}
```
I am not able to get a type for a pet. In other words, I can not do this
```
async getUserPet(): Promise<PetType>) {
    const { data } = ...query(USER_QUERY, {}).toPromise();

    return data.me.pet;
}
```

Right?

**Reply by JoviDeCroock** [reply]
You can use a generic to `query` as shown in the example - which is a TypeScript feature and is _distinct_ from TypedDoc

**Reply by jangot** [reply]
Do you mean I need to describe the types manually?

**Reply by jangot** [reply]
Sorry I really don't understand. When I use codegen I had all types for all my data structure without any additional effort. And I don't understand how to have same result.

**Reply by JoviDeCroock** [reply]
When you codegen for React/... yes there are integrations for that, for pure Node you can generate the base-types that are used in those functions _or_ generate typed-document nodes which imho is much safer and more elegant

**Reply by jangot** [reply]
Yes, my be typed-document is more elegant but I need to start from the more traditional approach.
And I can not find a description or an example of how to generate the base-types.
My be you can put here a small example.
